API: Fix structured dtype cast-safety, promotion, and comparison

seberg opened the pull request:
This PR replaces the old gh-15509 implementing proper type promotion
for structured voids.  It further fixes the casting safety to consider
casts with equivalent field number and matching order as "safe"
and if the names, titles, and offsets match as "equiv".

The change perculates into the void comparison, and since it fixes
the order, it removes the current FutureWarning there as well.

This addresses https://github.com/liberfa/pyerfa/issues/77
and replaces gh-15509 (the implementation has changed too much).

Fixes gh-15494  (and probably a few more)

Co-authored-by: Allan Haldane <allan.haldane@gmail.com>

---

Ping @ahaldane, @mhvk.  The tests will fail right now.  And we have to decide a few things, and probably one more change that is necessary here:

- [x] The tests fail, because the old PR included a test which ensured that the promotion result is always a packed struct.  I think this makes sense, but we further should ensure the promotion result is always native byte order (the old code probably did this also).  The new code doesn't call "promotion" because both descriptors are identical.  This calls `ensure_nbo`.  My plan for `ensure_nbo(descr)` is to actually call a new `DType.ensure_canonical(descr)` (for native dtypes, this will just do the same as `ensure_nbo`) but for structured ones it would: Recursively call `ensure_canonical` and drop offsets.  (Fixing the test failure)  *This should be simple, but requires adding the new slot to the DType, I will probably just do this soon.*  
- [x] I removed the `FutureWarning` from the void comparison path.  This means we have to decide wether void comparisons should ever return an array of `False` if the types they include mismatch.  Or if they should raise an error.  Note that most field mismatches will still raise a `FutureWarning` (since it will call `self[field1_self] == other[field1_other]` which will give a warning when the types are not comparable.  This is thus only about the explicit error messages below.  (Or alternatively, keep the warning for now, even if it is very strange for dtypes that can promote to not also compare.)
- [x] We need to improve test coverage (at least in all those points that coverage points too)

**As such this is currently draft, although all changes should be good, there are some things missing.  But I decided to have a look since gh-15509 is outdated now and it might be tricky to resurrect it without knowing the new layout well.**

**EDIT: Currently based off gh-19346**

seberg committed:
API: Fix structured dtype cast-safety, promotion, and comparison

This PR replaces the old gh-15509 implementing proper type promotion
for structured voids.  It further fixes the casting safety to consider
casts with equivalent field number and matching order as "safe"
and if the names, titles, and offsets match as "equiv".

The change perculates into the void comparison, and since it fixes
the order, it removes the current FutureWarning there as well.

This addresses https://github.com/liberfa/pyerfa/issues/77
and replaces gh-15509 (the implementation has changed too much).

Fixes gh-15494  (and probably a few more)

Co-authored-by: Allan Haldane <allan.haldane@gmail.com>

mhvk commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -3162,10 +3162,25 @@ can_cast_fields_safety(PyArray_Descr *from, PyArray_Descr *to)
     if (field_count != PyTuple_Size(to->names)) {
-        /* TODO: This should be rejected! */
-        return NPY_UNSAFE_CASTING;
+        return -1;
```
Yay! (as long as `np.zeros((), dtype)` continues to work!)

mhvk commented on numpy/core/tests/test_dtype.py:
```diff
@@ -151,11 +151,11 @@ def test_field_order_equality(self):
         assert_equal(x == y, False)
-        # But it is currently an equivalent cast:
-        assert np.can_cast(x, y, casting="equiv")
+        # This is an safe cast (not equiv) due to the different names:
```
This is a really tricky example, as the way it was written, it described exactly the same bytes in memory, but just with a different order. With the change to use field order, I guess it makes sense that, as written, it would become "unsafe" (integer to float and vice versa). Might it be good to keep that as an example?

And then also insert the new example, where, because it is by field order, it is "safe".

mhvk reviewed (commented):
This all looks quite clear, but there is one point that probably needs to be clarified. Currently, the docstring for `np.promote_types` states:
```
Returns the data type with the smallest size and smallest scalar
kind to which both ``type1`` and ``type2`` may be safely cast.
```
But for structured dtypes, this PR insists names match, which implies `casting='equiv'`.

This does seem intentional, but I think it is good to ensure it is the right decision. Should `np.promote_types` correspond to `casting='safe'` instead? Looking back at #15494, which concerns the concatenation of two arrays with the same dtype but different names, it is not obvious to me that should be disallowed (though I'd definitely expect the output dtype have names from the dtype of the first array). 

Note that I'm far from sure what the right answer is. But if it is disallowed, we should clarify this in the docs for `promote_types`. And more generally it might make sense to have some overview of when and how field names are used!

p.s. Probably too unrelated: should the following work without a copy?
```
a = np.array([(1., 2.), (3., 4.)], dtype=[('a', 'f8'), ('b', 'f8')])
a.astype([('b', 'f8'), ('a', 'f8')], copy=False)
```

mhvk commented:
Hmm, it seems the answer to my question about the behaviour of `promote_types` was already given by @ahaldane in https://github.com/numpy/numpy/issues/15494#issuecomment-581398363:
> It seems like we want for two dtypes with different field names but same field dtypes:
> * can be cast to each other
> * cannot be promoted to a common type unambiguously
>
> So can_cast should succeed, but promote_types should fail. Does that make sense?

I think "unambigously" is key here - with name mismatches, one has to pick, which means the result is ambiguous (and it breaks the symmetry promise of `np.promote_types`). So, this PR should include an update to the `np.promote_types` docstring...

@seberg - could you also include in this PR the changelog entry and doc updates to `structured_arrays.py` from #15509? They help understand what is being done!

On the larger scheme, mostly out of curiosity: is the goal that eventually the `DType` classes themselves can tell how promotion/casting should be done? I.e., that if one wanted to have a pure by-order or by-name dtype, one could implemented this?

seberg commented:
Yes, sorry, I should have copied over the doc changes/change log from the other PR and expanded on that, its a bit too raw right now here!   (And I thought adding the `ensure_canonical` is a bit more straight forward.)

---

About the "safe" casting... My opinion is that the documentation should be updated.  And I can probably do that even ignoring this PR.  I have given up on the idea that the common dtype (promotion) *derives* from casting.  I honestly think its a fallacy.  The two are related, but distinct.

Rather the properties of `P = promote_types(A, B)` are:

* Both `A` and `B` must (should?) safely cast to `P`.  (It seems surprising if we allow violating this.  But arguable we already *do* with `promote_types(int64, float64) -> float64` which is not *really* a safe cast.  The main reason we call it a safe cast is just to make promotion work – because it previously relied on safe casting.)
* We do *not* guarantee this is the "smallest".  This is *typical* and guaranteed for vanilla NumPy, but not a generic rule!
  * If a user adds an `int24` both `uint16` and `int16` can safely cast to it.  But `promote_types(uint16, int16)` *must* continue to return `int32`!
* If `A` can "equiv" cast to `B`, then I suppose we have some strong rules:
  * `P` must exist.
  * `P` can "equiv" cast to `B`
  * `B` can "equiv" cast to `P`
  * (`P` may be `B`, or `B.copy()` or `B.ensure_canonical()` – the last one is a byte-swap or removal offsets/padding.)

Ensuring this is the job of the type promotion (i.e. the DType implementer has to take care).  But in that definition type promotion can be more strict.

---

> p.s. Probably too unrelated: should the following work without a copy?
> ```
> a = np.array([(1., 2.), (3., 4.)], dtype=[('a', 'f8'), ('b', 'f8')])
> a.astype([('b', 'f8'), ('a', 'f8')], copy=False)
> ```

It is absolutely related!  But the fix/change itself is mostly orthongal.  We have to modify `arr.astype()` to use the following logic:
```
casting = get_casting_safety(self.dtype, new_dtype)
if casting == "no":
    # the current branch, should maybe use `self.dtype is new_dtype`, though!
    return self
if casting & cast_is_view:
    # new branch!
    return self.view(new_dype)
else:
    # create new array and cast
```
The changes here are designed so that that if we do that change, your example will also be no-copy.

---

> is the goal that eventually the DType classes themselves can tell how promotion/casting should be done? I.e., that if one wanted to have a pure by-order or by-name dtype, one could implemented this?

This is already the case.  But you don't have the public API to "recurse" into arbitrary dtypes well (e.g. to write the casting functions).  So you could write such a new DType today, but you have to do it in NumPy.
Making enough of that public needs more careful though than most of the API, so I think it is longer term than general user DTypes.  On the other hand, I am also not sure that it is actually problematic to just develop those as part of NumPy!  "Dynamically" structured DTypes are maybe just special enough :).

mhvk commented:
That really helps! I also like your proposed logic for `astype`, using the (hidden) viewable casting.

p.s. I think `ensure_canonical` is a good idea.

seberg commented on numpy/core/_internal.py:
```diff
@@ -433,6 +433,37 @@ def _copy_fields(ary):
+            "NumPy currently does not support promotion with field titles.")
+
+    return dtype([(name, promote_types(dt1[name], dt2[name]))
+                  for name in dt1.names])
```
Jeesh, I realized that this is actually incorrect.  What is missing is something like `align=dt1.isalignedstruct or dt2.isalignedstruct`.

Is that what we want? Preserve alignment if one is aligned, or drop alignment when one is not aligned?

mhvk replied:
I agree with your choice of preserving alignment when one is aligned.

seberg committed:
DOC: Add a long release-note about structured casts and promotion

mattip commented on doc/release/upcoming_changes/19226.compatibility.rst:
```diff
@@ -0,0 +1,85 @@
+
+The main important change here is that name mismatches are now
+considered "safe" casts.
+
```
This seems too long for a release note. Could it instead summarize the change and point to some more permanent documentation that discusses the details of casting for structured arrays?

mhvk replied:
:+1: - but the text is very good, so do use it! Including in `np.result_types` and `np.promote_types` docstrings.

rossbar replied:
This was at least partially addressed in 14601c and 9967058, but should be checked to see if those changes are sufficient!

mattip commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -897,22 +897,6 @@ promote_types(PyArray_Descr *type1, PyArray_Descr *type2,
-        return PyArray_DescrNewByteorder(type, NPY_NATIVE);
-    }
-}
-
```
This is no longer needed for non-structured arrays?

seberg replied:
It is just moved, but maybe I should limit the change a bit (since this is perfectly fine for most NumPy builtin dtypes).  In any case, it is actually a change of one of the previous PRs that would have to go in first.

mhvk replied:
See #19345

seberg committed:
BUG,MAINT: Never use negative offset and tighten promote-types

Negative offsets in structured dtypes give indirect access outside
of the original data in principle, thus we should always reject
them (this is as of now practically unused in NumPy).

Further, `np.result_type` and `np.promote_types` should behave
the same as canonicalization for twice the same input (at least
mostly, metadata is a rare exception).

This was not the case, because promote-types had a fast path that
would return the wrong thing for structured dtypes.
Thus, reject them from the fast-path, and limit it to builtin types
for now.

seberg commented:
Code wise, this should be mostly ready now.  I narrowed the special case in `promote_types` (I have not much love for that, but I have no better idea).  I need to reorganize the docs with the release notes and review the old comments though.

mhvk commented:
@seberg - sounds good. Probably easiest to review once the doc changes are ready too...

seberg committed:
WIP: docs... Need to start a whole new "promotion" probably based on some stuff in the NEP 50 draft...

seberg committed:
fixup promote types

seberg committed:
MAINT,ENH: Adjust metadata preservation, preserving it for identical structured dtypes

Identical means here that the promotion is between the same dtype *and* promoting it
does not modify any of the included dtypes *and* does not modify the offset layout
or itemsize.

seberg committed:
TST: Fortify tests for improved struct dtype promotion and comparison

seberg committed:
BUG: Add missing return NULL in void common instance

seberg committed:
TST: Fixup slow metadata promotion tests to match new behaviour

seberg committed:
TST: Add additional (basic) failure tests for void comparisons

Also modify the error message to remove the "may return an array
of False note".  That is fathomable, but if we really want to go
there, maybe it should just get a new FutureWarning...

rossbar committed:
Rm preamble from release note.

Most of the removed info is already in the structured array dtype
doc or later in the release note - leave high-level summary in place.

rossbar committed:
DOC: mv details from release note to user guide.

seberg committed:
BUG: Failed comparison should be TypeError and fix title mismatch test

seberg committed:
TST: Fix byte-swap test for big-endian machines

One of the following test is mainly interesting on little endian machines,
but that seems OK, the tests definitely run on somewhere :)

seberg commented:
@ahaldane just a ping in case you have time to have a look on the logic, or whether the tests make sense.  (I think it would be good to include this in the upcoming release and branching is looming, to finally get us all the way on "order", rather than "names" with structured dtypes.)

rossbar reviewed (approved):
This LGTM, though I mostly focused on the Python components + tests!

seberg committed:
DOC: Fixup test and general doc strings

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

mattip commented:
The linting "errors" are fine to ignore. Thanks @seberg

Kongpcmail commented:
> This PR replaces the old [gh-15509](https://github.com/numpy/numpy/pull/15509) implementing proper type promotion for structured voids. It further fixes the casting safety to consider casts with equivalent field number and matching order as "safe" and if the names, titles, and offsets match as "equiv".
> 
> The change perculates into the void comparison, and since it fixes the order, it removes the current FutureWarning there as well.
> 
> This addresses [liberfa/pyerfa#77](https://github.com/liberfa/pyerfa/issues/77) and replaces [gh-15509](https://github.com/numpy/numpy/pull/15509) (the implementation has changed too much).
> 
> Fixes [gh-15494](https://github.com/numpy/numpy/issues/15494) (and probably a few more)
> 
> Co-authored-by: Allan Haldane [allan.haldane@gmail.com](mailto:allan.haldane@gmail.com)
> 
> Ping @ahaldane, @mhvk. The tests will fail right now. And we have to decide a few things, and probably one more change that is necessary here:
> 
> * [x]  The tests fail, because the old PR included a test which ensured that the promotion result is always a packed struct.  I think this makes sense, but we further should ensure the promotion result is always native byte order (the old code probably did this also).  The new code doesn't call "promotion" because both descriptors are identical.  This calls `ensure_nbo`.  My plan for `ensure_nbo(descr)` is to actually call a new `DType.ensure_canonical(descr)` (for native dtypes, this will just do the same as `ensure_nbo`) but for structured ones it would: Recursively call `ensure_canonical` and drop offsets.  (Fixing the test failure)  _This should be simple, but requires adding the new slot to the DType, I will probably just do this soon._
> * [x]  I removed the `FutureWarning` from the void comparison path.  This means we have to decide wether void comparisons should ever return an array of `False` if the types they include mismatch.  Or if they should raise an error.  Note that most field mismatches will still raise a `FutureWarning` (since it will call `self[field1_self] == other[field1_other]` which will give a warning when the types are not comparable.  This is thus only about the explicit error messages below.  (Or alternatively, keep the warning for now, even if it is very strange for dtypes that can promote to not also compare.)
> * [x]  We need to improve test coverage (at least in all those points that coverage points too)
> 
> **As such this is currently draft, although all changes should be good, there are some things missing. But I decided to have a look since [gh-15509](https://github.com/numpy/numpy/pull/15509) is outdated now and it might be tricky to resurrect it without knowing the new layout well.**
> 
> **EDIT: Currently based off [gh-19346](https://github.com/numpy/numpy/pull/19346)**
